### PR TITLE
[Feat] Add Sequence Parallelism (USP) support for HunyuanVideo 1.5 transformer

### PR DESCRIPTION
**Pull request by daixinning**
## Purpose

This PR adds Sequence Parallelism (Ulysses SP) support for HunyuanVideo 1.5 (T2V and I2V), enabling distributed inference across multiple devices by sharding the video token sequence across SP ranks.

HunyuanVideo 1.5 uses a dual-stream (joint) attention architecture where video and text tokens are processed in the same attention layer. This requires special handling for SP: video tokens are sequence-sharded via `_sp_plan` hooks, while text (encoder) tokens are passed as `joint_*` metadata and head-sliced within the Ulysses attention layer.

### Changes

**`hunyuan_video_15_transformer.py`**:
- Add `_sp_plan` to `HunyuanVideo15Transformer3DModel`:
  - `rope`: split RoPE embeddings (freqs_cos/sin) along seq dim with auto_pad
  - `transformer_blocks.0`: split `hidden_states` along seq dim at first block entry
  - `proj_out`: gather output along seq dim after final projection
- Add SP-aware attention path in `HunyuanVideo15Attention`:
  - When SP is active, encoder Q/K/V are passed via `joint_query/key/value` with `joint_strategy="rear"`, allowing Ulysses to head-slice encoder tokens and all-to-all video tokens independently
  - When SP is inactive, the original concat path is preserved (no behavior change)

## Test Plan

Tested with offline text-to-video generation using `examples/offline_inference/text_to_video/text_to_video.py`:

```bash
python examples/offline_inference/text_to_video/text_to_video.py \
    --model HunyuanVideo-1.5-Diffusers-480p_t2v \
    --prompt "A little girl wearing a straw hat runs through a summer meadow full of wildflowers." \
    --num-frames 61 --num-inference-steps 50 --seed 42 \
    --tensor-parallel-size 1 \
    --ulysses-degree 2 \
    --ring-degree 1 \
    --cfg-parallel-size 2 \
    --vae-patch-parallel-size 4 \
    --vae-use-tiling \
    --output t2v_480p.mp4
```

## Test Result

Tested on 4x Ascend NPUs, HunyuanVideo-1.5-Diffusers-480p_t2v, 480p resolution, 61 frames, 50 denoising steps.

| Configuration | Denoising (50 steps) | Per-step | Decoding | Total Pipeline |
|---|---|---|---|---|
| Baseline: tp=2, ulysses=1, cfg=2 | 149,651 ms | 2,993 ms | 5,580 ms | 155,443 ms |
| **SP enabled: tp=1, ulysses=2, cfg=2** | **131,358 ms** | **2,627 ms** | **5,641 ms** | **137,202 ms** |

SP (ulysses_degree=2) reduces denoising time by **12.2%** compared to the TP=2 baseline, with a **11.7%** reduction in total pipeline wall time.

Logs confirm SP hooks are correctly applied:
```
Applying sequence parallelism to HunyuanVideo15Transformer3DModel (transformer) (sp_size=2, mode=ulysses, ulysses=2, ring=1)
```

https://github.com/user-attachments/assets/c6574cf3-0291-411f-9938-a92293ff4efc

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR
- [x] The test plan with test scripts and commands
- [x] The test results with before/after comparison
- [ ] (Optional) Documentation update
- [ ] (Optional) Release notes update
</details>

### CODE REVIEW

File: vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py [hunk 1]
```diff
@@ -430,22 +432,41 @@ def forward(
...
-        hidden_states = self.attn(query, key, value, attn_metadata)
+        if use_sp and encoder_hidden_states is not None:
+            attn_metadata = AttentionMetadata(
+                joint_query=encoder_query,
+                joint_key=encoder_key,
+                joint_value=encoder_value,
+                joint_strategy="rear",
+            )
+            if attention_mask is not None:
+                attn_metadata.joint_attn_mask = attention_mask.bool()
```
[the remark this answers is not in the document]
Comment (daixinning) [reply vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py:451]: Fix (already applied in latest push).

**Comment by hsliuustc0106** [comment]
what's the best parallel strategy within 2 devices? can you provide device profiling to check the details?

**Comment by daixinning** [comment]
> what's the best parallel strategy within 2 devices? can you provide device profiling to check the details?

Thanks for the question! Currently I only have access to a 4-device (4x Ascend NPU) environment and don't have a 2-device setup available for profiling. The test configuration in this PR uses ulysses_degree=2, cfg_parallel_size=2 across 4 devices.

File: vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py [hunk 2]
```diff
@@ -395,6 +398,7 @@ def forward(
         encoder_hidden_states: torch.Tensor | None = None,
         attention_mask: torch.Tensor | None = None,
         image_rotary_emb: tuple[torch.Tensor, torch.Tensor] | None = None,
+        hidden_states_mask: torch.Tensor | None = None,
     ) -> tuple[torch.Tensor, torch.Tensor]:
```
[the remark this answers is not in the document]
Comment (daixinning) [reply vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py]: Fixed. Updated the return type annotation to torch.Tensor | tuple[torch.Tensor, torch.Tensor] for both HunyuanVideo15Attention.forward and HunyuanVideo15TransformerBlock.forward.

File: vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py [hunk 3]
```diff
@@ -717,13 +755,29 @@ def forward(
...
+            padded_seq_len = ctx.sp_original_seq_len + ctx.sp_padding_size
+            hidden_states_mask = torch.ones(
+                batch_size,
+                padded_seq_len,
+                dtype=torch.bool,
+                device=hidden_states.device,
+            )
+            hidden_states_mask[:, ctx.sp_original_seq_len :] = False
+            if hidden_states_mask.all():
+                hidden_states_mask = None
```
[the remark this answers is not in the document]
Comment (daixinning) [reply vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py:780]: Agreed, removed. The hidden_states_mask is only constructed when sp_padding_size > 0 and the tail is explicitly set to False, so .all() can never be True in this branch.

File: vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py [hunk 4]
```diff
@@ -430,22 +434,43 @@ def forward(
...
+                joint_query=encoder_query,
+                joint_key=encoder_key,
+                joint_value=encoder_value,
+                joint_strategy="rear",
+            )
+            if attention_mask is not None:
+                attn_metadata.joint_attn_mask = attention_mask.bool()
+            if hidden_states_mask is not None:
+                attn_metadata.attn_mask = hidden_states_mask
+            hidden_states = self.attn(query, key, value, attn_metadata)
```
[the remark this answers is not in the document]
Comment (daixinning) [reply vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py]: Good catch. Fixed — replaced the use_sp variable (based on get_sequence_parallel_world_size() > 1) with get_forward_context().sp_active, so the joint-metadata path is only taken when SP is actually active in the current forward context. This avoids the potential mask shape mismatch when SP world_size > 1 but sp_active is false.

File: vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py [hunk 5]
```diff
@@ -717,13 +748,27 @@ def forward(
...
+        if ctx.sp_original_seq_len is not None and ctx.sp_padding_size > 0:
+            padded_seq_len = ctx.sp_original_seq_len + ctx.sp_padding_size
+            hidden_states_mask = torch.ones(
+                batch_size,
+                padded_seq_len,
+                dtype=torch.bool,
+                device=hidden_states.device,
+            )
+            hidden_states_mask[:, ctx.sp_original_seq_len :] = False
+
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py:777]: Nit: this tensor is re-allocated every denoising step but its content is identical across steps. You could stash it on the first call and reuse, though given it's a small bool mask the overhead is negligible — up to you.

File: vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py [hunk 6]
```diff
@@ -430,22 +433,37 @@ def forward(
...
+            )
+            if attention_mask is not None:
+                attn_metadata.joint_attn_mask = attention_mask.bool()
+            if hidden_states_mask is not None:
+                attn_metadata.attn_mask = hidden_states_mask
+            hidden_states = self.attn(query, key, value, attn_metadata)
+        else:
+            if encoder_hidden_states is not None:
+                query = torch.cat([query, encoder_query], dim=1)
+                key = torch.cat([key, encoder_key], dim=1)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py:458]: Worth adding a brief inline comment here explaining *why* the encoder tokens go via `joint_*` metadata instead of the concat path — i.e. Ulysses needs to head-slice them separately from the all-to-all'd video tokens. Would help future readers understand the split.

**Review by lishunyang12 — COMMENTED** [review]
left a couple of minor comments, overall looks good

File: vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py (hunk 5 above)
Comment (daixinning) [reply vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py:777]: Good suggestion, applied. The mask is now cached as `self._cached_sp_mask` on first construction and reused across subsequent denoising steps. While the overhead is indeed small, caching avoids the repeated allocation and keeps the code cleaner.

File: vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py (hunk 6 above)
Comment (daixinning) [reply vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py:458]: Makes sense, added a comment above the `attn_metadata` construction block explaining that under Ulysses SP, encoder tokens are passed via `joint_*` metadata so they can be head-sliced separately from the all-to-all'd video tokens in `UlyssesParallelAttention`.

File: vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py [hunk 7]
```diff
@@ -717,13 +751,29 @@ def forward(
         encoder_hidden_states = torch.stack(new_encoder_hidden_states)
         encoder_attention_mask = torch.stack(new_encoder_attention_mask)
 
+        # Create explicit attn_mask for image tokens when SP auto_pad is active.
+        # Cache the mask since it is identical across denoising steps.
+        ctx = get_forward_context()
+        hidden_states_mask = getattr(self, "_cached_sp_mask", None)
+        if hidden_states_mask is None and ctx.sp_original_seq_len is not None and ctx.sp_padding_size > 0:
+            padded_seq_len = ctx.sp_original_seq_len + ctx.sp_padding_size
```
Comment (wtomin) [inline vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py:769]: `self._cached_sp_mask` maybe outdated if new batches use different resolution x num_frames. Since you just need to reuse this cached mask for each diffusion step, you may need to manually reset the cached mask to None in the pipeline, after all diffusion step are finished.

Comment (daixinning) [reply vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py:769]: Good catch. On further analysis, the mask is stored on the transformer singleton so it persists across requests, not just across denoising steps. Since there is no safe per-request cache key available at the transformer level, and the mask construction overhead is negligible (torch.ones + one slice assignment), we removed the cache entirely rather than trying to invalidate it. Fixed in a65b272d.

File: vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py (hunk 5 above)
Comment (daixinning) [reply vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py:777]: Update: we ended up reverting the cache (a65b272d). The mask is stored on the transformer singleton, so it persists across requests — not just across denoising steps within one request. Since the transformer has no access to a per-request identifier, there is no safe way to invalidate the cache between requests. Given the construction overhead is negligible as you noted, removing the cache is cleaner than trying to manage its lifetime.

**Comment by lishunyang12** [comment]
Show visual output

**Comment by hsliuustc0106** [comment]
@Bounty-hunter PTAL

**Comment by daixinning** [comment]
Here are the visual outputs for the T2V inference on 4x Ascend NPU (832×480, 33 frames, 50 steps).

**USP=2 + CFG parallel=2** (ulysses_degree=2, cfg_parallel_size=2):
- Denoising: 68.0 s, Total: 81.1 s

https://github.com/user-attachments/assets/0169fcff-ed43-4d70-a1dc-3280af212dbc

**USP=4** (ulysses_degree=4, cfg_parallel_size=1):
- Denoising: 83.6 s, Total: 96.8 s

https://github.com/user-attachments/assets/f0f75015-3246-4e1a-8191-a72fe2d9be56

Both configurations produce correct visual output. USP=2+CFG=2 is faster on 4 devices since CFG parallel has lower communication overhead than SP for this sequence length.

@lishunyang12

**Review by lishunyang12 — APPROVED** [review]
## Review: Sequence Parallelism (USP) support for HunyuanVideo 1.5

I reviewed the full diff and traced the data flow through the Ulysses attention infrastructure. The changes are correct and well-structured.

### Correctness analysis

**`_sp_plan` configuration**: RoPE uses `split_dim=0, expected_dims=2` which correctly matches the `(seq, dim)` shape of `freqs_cos`/`freqs_sin`. `hidden_states` split at `transformer_blocks.0` with `split_dim=1, expected_dims=3` is correct for `(B, S, C)`. Gather at `proj_out` reverses this. All entries use `auto_pad=True` as needed.

**Joint attention under SP**: The approach of passing encoder Q/K/V via `joint_query/key/value` with `joint_strategy="rear"` is the correct pattern for dual-stream models with Ulysses SP. I verified that:
- `UlyssesParallelAttention.post_attention()` correctly splits video and encoder portions, applies AllToAll on video and AllGather on encoder, then recombines in the same order.
- The `split_with_sizes` at the end of `HunyuanVideo15Attention.forward` correctly separates them back since the Ulysses output preserves the `[video | encoder]` concatenation order.

**Mask handling**: The `hidden_states_mask` with `padded_seq_len` matches the full sequence dimension after Ulysses AllToAll restores it. The mask is set as `attn_metadata.attn_mask` (video padding), while the existing encoder attention mask is set as `attn_metadata.joint_attn_mask`. The Ulysses layer combines them at lines 400-404 via `cat([attn_mask, joint_attn_mask], dim=1)` for `rear` strategy, producing the correct full query-length mask.

**Non-SP path preserved**: The `else` branch faithfully reproduces the original concatenation logic with no behavior change when SP is inactive. Good.

### Minor suggestions (non-blocking)

1. **Missing `hidden_states_mask.all()` guard**: The Wan model includes `if hidden_states_mask is not None and hidden_states_mask.all(): hidden_states_mask = None` as defensive cleanup. While the `sp_padding_size > 0` guard in this PR makes it impossible for the mask to be all-True, adding the same guard would be consistent with the codebase convention and protect against future refactors.

2. **`get_forward_context()` called per-block**: `ctx = get_forward_context()` is called inside `HunyuanVideo15Attention.forward` on every block invocation. This is fine for correctness (it's a lightweight lookup), but you could consider caching `ctx.sp_active` as a local variable at the block level if profiling ever shows overhead with 54 blocks.

Overall this is a clean, well-tested integration that follows the established SP patterns in the codebase. The 12% denoising speedup is a solid result.

LGTM.

**Comment by daixinning** [comment]
Two follow-up commits added after the approval:

5ac9925 Fix: support cache_dit with USP

When cache_dit is stacked on top of USP, CachedBlocks.forward saves original_hidden_states at the block entry boundary — which was transformer_blocks.0 in the original
_sp_plan. At that point the hidden states had not yet been sharded, so CacheDiT was caching full-sequence tensors while the actual forward pass operated on per-rank shards,
causing a shape mismatch.

Fix: move the SP scatter of hidden_states from _sp_plan to after x_embedder (before transformer_blocks), so CacheDiT always sees already-sharded tensors. Also adds
enable_cache_for_hunyuan_video_15 in cache_dit_backend.py and registers it for both HunyuanVideo15Pipeline and HunyuanVideo15I2VPipeline.

LPIPS validation (29 frames, AlexNet): mean = 0.0467 between baseline and cache_dit+USP outputs — no visible quality degradation.

5c43134 Fix: add hidden_states_mask.all() guard (suggestion 1 from @lishunyang12)

Added the defensive if hidden_states_mask.all(): hidden_states_mask = None cleanup after mask construction, consistent with the pattern used in the Wan model.

**Comment by lishunyang12** [comment]
@wtomin PTAL

**Comment by hsliuustc0106** [comment]
resolve conflicts please
